### PR TITLE
discovery: add more tests

### LINKED ISSUE #4565: Make the Discovery throtling configurable to allow easier testing.
https://github.com/prometheus/prometheus/blob/75bd348135019c6bf5bee4350a339964e62f2d8a/discovery/manager.go#L134

Sending the discovery updates are hard coded to throttle at 5 seconds. 
This makes creaating test for these parts of the code needing to wait 5 seconds so not good for unit testing. We need to make this somehow configurable so that we can adjust when creating unit tests.

With this change to the same PR  can also add a test for:

 - blocked or no updates when the receiver's channel is full - https://github.com/prometheus/prometheus/pull/4523#issuecomment-415967435
 - send updates only when there was an actual change in the targets - https://github.com/prometheus/prometheus/pull/4545

### PR DESCRIPTION
**Pull request by simonpasquier**
Closes #4565

When I run this PR without #4556, it fails as expected.

### CODE REVIEW

File: discovery/manager_test.go [hunk 1]
```diff
@@ -868,40 +885,289 @@ scrape_configs:
...
+								Source:  "tg2",
+								Targets: []model.LabelSet{{"__instance__": "2"}},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			title: "The receiver gets an update when a target group is gone",
```
Comment (krasi-georgiev) [inline discovery/manager_test.go]: doesn't this and all the tests below already duplicate with the tests in `TestTargetUpdatesOrder` ?

**Comment by simonpasquier** [comment]
Thanks for the review @krasi-georgiev 

I've added an explicit test (`TestCoordinationWithEmptyProvider`) when the provider closes the update channel although it was implicitly tested by `TestTargetSetRecreatesTargetGroupsEveryRun`. I've also modified the tests in `TestCoordinationWithReceiver` to fail if they didn't receive anything after 5 seconds while something is still expected.

> doesn't this and all the tests below already duplicate with the tests in TestTargetUpdatesOrder ?

Somehow yes but this test is more end-to-end since it reads from the SD manager update channel while `TestTargetUpdatesOrder` bypasses `updater()`. I should probably rewrite `TestTargetUpdatesOrder` to check the output of the update channel. WDYT?

Finally regarding
> https://github.com/prometheus/prometheus/blob/068eaa5dbfce6c08f3d05d3d3c0bfd96267cfed2/discovery/manager.go#L204-L207

This is covered by the `"Receiver should get updates even when the channel is blocked"` test. See https://github.com/prometheus/prometheus/pull/4582/files?utf8=%E2%9C%93&diff=unified#diff-f7fd37c9f5b407b3f4f1bd3ef08c624dR902

**Comment by krasi-georgiev** [comment]
> I've added an explicit test (TestCoordinationWithEmptyProvider) when the provider closes the update channel although it was implicitly tested by TestTargetSetRecreatesTargetGroupsEveryRun. I've also modified the tests in TestCoordinationWithReceiver to fail if they didn't receive anything after 5 seconds while something is still expected.

I had something else in mind.
Start a provider ->  send some updates -> close the channel before the discovery manager's `updatert` 
this ensures that any updates before `updatert` are sent even when the provider closes its channel

**Comment by krasi-georgiev** [comment]
>>  doesn't this and all the tests below already duplicate with the tests in TestTargetUpdatesOrder ?

> Somehow yes but this test is more end-to-end since it reads from the SD manager update channel while TestTargetUpdatesOrder bypasses updater(). I should probably rewrite TestTargetUpdatesOrder to check the output of the update channel. WDYT?

yeah I would say that these test fit better in the `TestTargetUpdatesOrder` so lets refactor.

File: discovery/manager_test.go [hunk 2]
```diff
@@ -868,40 +885,354 @@ scrape_configs:
...
+		tgs   map[string][]*targetgroup.Group
+	}
+
+	testCases := []struct {
+		title     string
+		providers map[string]Discoverer
+		expected  []expect
+	}{
+		{
+			title: "Receiver should get an empty map even when the provider sends nothing and closes the channel",
```
Comment (krasi-georgiev) [inline discovery/manager_test.go]: why do we need to test for this? Was there a bug for this?

File: discovery/manager_test.go [hunk 3]
```diff
@@ -868,40 +885,354 @@ scrape_configs:
...
+								Source:  "tg2",
+								Targets: []model.LabelSet{{"__instance__": "2"}},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			title: "Receiver should get updates even when the channel is blocked",
```
Comment (krasi-georgiev) [inline discovery/manager_test.go]: What about:
"Receiver should get all updates even when the channel was blocked"

File: discovery/manager_test.go [hunk 4]
```diff
@@ -868,40 +885,354 @@ scrape_configs:
...
+				"empty": &onceProvider{},
+			},
+			expected: []expect{
+				{
+					tgs: map[string][]*targetgroup.Group{},
+				},
+			},
+		},
+		{
+			title: "Receiver should get updates even when one provider closes its channel",
```
Comment (krasi-georgiev) [inline discovery/manager_test.go]: What about: "Receiver should get all updates even when one provider closes its channel"

File: discovery/manager_test.go (hunk 2 above)
Comment (simonpasquier) [reply discovery/manager_test.go]: Not really. This is the result of some exploratory testing but agreed that it isn't necessary.

**Comment by simonpasquier** [comment]
BTW I've found something that could be optimized in the manager: there are as many `updater()` goroutines as discoverers meaning that the updates can be sent to the channel more often than every 5 seconds when you have many discoverers.

It is trivial to fix it but I wonder if I miss something here? I've got a local branch that moves the sending of the updates to the channel from `updater()` to a single goroutine and it passes the new tests.

**Comment by krasi-georgiev** [comment]
ooh yeah I think we discussed this somewhere.
Lets merge this one to keep it small and can open a PR for that optimisation as well.

**Comment by simonpasquier** [comment]
> Lets merge this one to keep it small

Sure even though this PR isn't really small already :smirk:

**Comment by simonpasquier** [comment]
I should have addressed all your comments. I'll refactor `TestTargetUpdatesOrder` with/after the optimization for dispatching the updates only once otherwise the tests aren't be reliable with multiple providers.

**Comment by krasi-georgiev** [comment]
Yep that is even better.

LGTM, Thanks!

**Comment by simonpasquier** [comment]
@gouthamve any opinion on this PR?

**Pull request merged**